File: 9-distinct.py
#'use strict';

#const distinct = (dataset) => {
#  const keys = new Set();
#  return dataset.filter((record) => {
#    const cols = Object.keys(record).sort();
#    const key = cols.map((field) => record[field]).join('\x00');
#    const has = keys.has(key);
#    if (!has) keys.add(key);
#    return !has;
#  });
# };

#// Usage

# const flights = [
#  { from: 'Kiev', to: 'Rome' },
#  { from: 'Kiev', to: 'Warsaw' },
#  { from: 'Dublin', to: 'Riga' },
#  { from: 'Riga', to: 'Dublin' },
#  { from: 'Kiev', to: 'Rome' },
#  { from: 'Cairo', to: 'Paris' },
# ];

# console.dir({ flights });
# const directions = distinct(flights);
# console.dir({ directions });

# ---------------------------------------------------------------

flights_from = ['Kiev', 'Kiev', 'Dublin', 'Riga', 'Kiev', 'Cairo']
flights_to = ['Rome', 'Warsaw', 'Riga', 'Dublin', 'Rome', 'Paris']
flights = [flights_from[x] + ' >> ' + flights_to[x] for x in range(len(flights_from))]
print('flights :', '\n', flights)
directions = list(dict.fromkeys(flights))
print('directions :', '\n', directions)

File: lecture_notes/08 collections/9-distinct.cpp
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <vector>

using namespace std;

using DataSetValue = map<string, string>;

void print_map(const DataSetValue& m) {
    cout << "Map: { ";
    for (const auto& kv : m) {
        cout << kv.first << " = " << kv.second << " ";
    }
    cout << "}" << endl;
}

void print_dataset(const vector<DataSetValue>& v) {
     cout << "[ " << endl;
     for (const auto& m : v) {
        cout << "\t";
        print_map(m);
     }
     cout << "]" << endl;
}

vector<DataSetValue> distinct(const vector<DataSetValue>& dataset) {
    map<string, DataSetValue> dataset_map;
    for (const auto& record : dataset) {
        string key;
        for (const auto& kv : record) {
            key += kv.second;
        }
        if (dataset_map.find(key) == dataset_map.end())
            dataset_map.insert({key, record});
    }

    size_t size = dataset_map.size();
    vector<DataSetValue> distinct_dataset(size);
    auto it = dataset_map.begin();
    for (size_t i = 0; i < size; ++i) {
        distinct_dataset[i] = (*it).second;
        it++;
    }

    return distinct_dataset;
}

int main() {
    vector<DataSetValue> flights {
        {{"from", "Kiev"}, {"to", "Rome"}},
        {{"from", "Kiev"}, {"to", "Warsaw"}},
        {{"from", "Dublin"}, {"to", "Riga"}},
        {{"from", "Riga"}, {"to", "Dublin"}},
        {{"from", "Kiev"}, {"to", "Rome"}},
        {{"from", "Cairo"}, {"to", "Paris"}}
    };
    print_dataset(flights);

    auto directions = distinct(flights);
    print_dataset(directions);

    return 0;
}
